# C9-dimensionality-reduction — Practice p05 — Solution

In [1]:
import numpy as np

SEED = 20260804
rng = np.random.default_rng(SEED)
B = rng.normal(0.0, 1.0, (150, 3))
M = np.array([[1.2, -0.4, 0.0, 0.8],
              [0.0, 0.9, 0.5, -0.3],
              [0.7, 0.0, -1.1, 0.2]], dtype=np.float64)
X = np.array([3.0, -1.0, 0.5, 2.0], dtype=np.float64) + B @ M + rng.normal(0.0, 0.25, (150, 4))

mu = X.mean(axis=0)
Xc = X - mu
col_gap = float(np.max(np.abs(Xc.mean(axis=0))))
var_cols = X.var(axis=0, ddof=1)
total_var = float(var_cols.sum())
_, s, _ = np.linalg.svd(Xc, full_matrices=False)
total_var_svd = float((s**2).sum() / (X.shape[0] - 1))
route_gap = float(abs(total_var - total_var_svd))
assert route_gap < 1e-9

print("mu:", mu)
print("column-mean gap:", col_gap)
print("per-column variances:", var_cols)
print("total variance by columns / spectrum:", total_var, total_var_svd)

mu: [ 2.8107966  -0.94850998  0.59437258  1.88516895]
column-mean gap: 3.3676765080296414e-16
per-column variances: [1.90309568 0.95426961 1.52162867 0.79225099]
total variance by columns / spectrum: 5.171244949844634 5.1712449498446365


Subtracting the row mean centers every feature.  The sum of the four sample variances is $5.1712449498$, and the singular-value identity gives the same total because $\|X_c\|_F^2=\sum_i\sigma_i^2$.

### Answer check

In [2]:
assert mu.shape == (4,)
assert Xc.shape == (150, 4)
assert col_gap < 1e-10
assert var_cols.shape == (4,)
assert np.isclose(total_var, 5.171244949844634, atol=1e-12, rtol=0)
assert route_gap < 1e-9